### PR TITLE
Prototype skip negative tests

### PR DESCRIPTION
**Pull request by jspaith**
Currently doing failure testing with `umock_c_negative_tests_fail_call` can be tedious because we need to explicitly specify the index of every call to skip.

I've added `umock_c_negative_tests_fail_call` to let the test caller skip calls that can't fail in a way that's not tied to index.  e.g.

```c
STRICT_EXPECTED(blah...)
umock_c_negative_tests_snapshot();
size_t count = umock_c_negative_tests_call_count();
for (size_t index = 0; index < count; index++) {
   if (umock_c_negative_tests_can_call_fail(index) == 0) continue;
   // else do negative test run
}
```

void functions automatically will be skipped, but additionally the product may ignore certain errors even on non-void.  Here the UT can specify this with 

```c
STRICT_EXPECTED_CALL(Blah()).SetReturn(NULL).CallCannotFail();
```

I've run this through gates a pre-check and built against C-utility and Azure IoT C SDK's UT to see no regressions.  Further pre-tetsing on some of C SDK UT's tests on both VS2017 & Ubuntu GCC with the functionality added works as expected..

### CODE REVIEW

**Comment by jspaith** [comment]
I'll squash the changes before checkin of course! - I like to backup to server.

File: devdoc/umockcallrecorder_requirements.md [hunk 1]
```diff
@@ -200,3 +201,24 @@ extern int umockcallrecorder_fail_call(UMOCKCALLRECORDER_HANDLE umock_call_recor
...
+
+
+## umockcallrecorder_can_call_fail
+
+```c
+extern int umockcallrecorder_can_call_fail(UMOCKCALLRECORDER_HANDLE umock_call_recorder, size_t index, int* can_call_fail);
+```
+
+
+**SRS_UMOCKCALLRECORDER_31_056: [** If umock_call_recorder or call_can_fail is NULL, umockcallrecorder_can_call_fail shall return a non-zero value. **]**
```
Comment (dcristoloveanu) [inline devdoc/umockcallrecorder_requirements.md]: >umock_call_recorder or call_can_fail [](start = 39, length = 36)

Consider splitting this in 2 requirements. It's a consider as I see the whole file does it the "old" way.

File: devdoc/umockcallrecorder_requirements.md [hunk 2]
```diff
@@ -200,3 +201,24 @@ extern int umockcallrecorder_fail_call(UMOCKCALLRECORDER_HANDLE umock_call_recor
...
+
+**SRS_UMOCKCALLRECORDER_31_057: [** If index is invalid, umockcallrecorder_can_call_fail shall return a non-zero value. **]**
```
Comment (dcristoloveanu) [inline devdoc/umockcallrecorder_requirements.md]: >invalid [](start = 48, length = 7)

Consider stating greater or equal to the current expected calls count (keyword being the fact that we evaluate it against the expected call count.

File: devdoc/umockcallrecorder_requirements.md [hunk 3]
```diff
@@ -200,3 +201,24 @@ extern int umockcallrecorder_fail_call(UMOCKCALLRECORDER_HANDLE umock_call_recor
...
+
+**SRS_UMOCKCALLRECORDER_31_057: [** umockcallrecorder_can_call_fail shall determine whether given call can fail or not by calling umockcall_set_fail_call. **]**
```
Comment (dcristoloveanu) [inline devdoc/umockcallrecorder_requirements.md]: >se [](start = 140, length = 2)

get

File: devdoc/umockcall_requirements.md [hunk 4]
```diff
@@ -23,6 +23,8 @@ umockcall is a module that encapsulates a umock call.
     extern int umockcall_get_fail_call(UMOCKCALL_HANDLE umockcall);
     extern int umockcall_set_ignore_all_calls(UMOCKCALL_HANDLE umockcall, int ignore_all_calls);
     extern int umockcall_get_ignore_all_calls(UMOCKCALL_HANDLE umockcall);
+    extern int umockcall_set_call_can_fail(UMOCKCALL_HANDLE umockcall, int call_can_fail);
```
Comment (dcristoloveanu) [inline devdoc/umockcall_requirements.md]: >extern [](start = 4, length = 6)

extern not needed. It is by default, so the standard says. Consider removing it for all functions, that's what I do these days when I hit them.

File: devdoc/umock_c_lib_requirements_with_ids.md [hunk 5]
```diff
@@ -893,6 +893,11 @@ TEST_FUNCTION(validate_argument_as_type_sample)
 }
 ```
 
+### call_cannot_fail_func_{name}  
+
+XX**SRS_UMOCK_C_LIB_31_209: [** `call_cannot_fail_func__{name}` call modifier shall record that when performing failure case run, this call should be skipped. **]**  
+X**SRS_UMOCK_C_LIB_31_210: [** If recording that the call cannot fail is unsuccessful, umock shall raise with the error code UMOCK_C_ERROR.  **]**
```
Comment (dcristoloveanu) [inline devdoc/umock_c_lib_requirements_with_ids.md]: Nice

File: devdoc/umock_c_lib_requirements_with_ids.md [hunk 6]
```diff
@@ -893,6 +893,11 @@ TEST_FUNCTION(validate_argument_as_type_sample)
 }
 ```
 
+### call_cannot_fail_func_{name}  
+
+XX**SRS_UMOCK_C_LIB_31_209: [** `call_cannot_fail_func__{name}` call modifier shall record that when performing failure case run, this call should be skipped. **]**
```
Comment (dcristoloveanu) [inline devdoc/umock_c_lib_requirements_with_ids.md]: >_fail_ [](start = 44, length = 6)

Can you put a space between the 2 lines for consistency with the rest of the doc? Thanks!

File: devdoc/umock_c_negative_tests_requirements.md [hunk 7]
```diff
@@ -13,6 +13,7 @@ umock_c_negative_tests is an addon to umock_c that exposes APIs used to automate
     extern void umock_c_negative_tests_reset(void);
     extern void umock_c_negative_tests_fail_call(size_t index);
     extern size_t umock_c_negative_tests_call_count(void);
+    extern int umock_c_negative_tests_can_call_fail(size_t index);
```
Comment (dcristoloveanu) [inline devdoc/umock_c_negative_tests_requirements.md]: >extern [](start = 4, length = 6)

Same consider about extern.

File: tests/umock_c_negt_ut/umock_c_negative_tests_ut.c [hunk 8]
```diff
@@ -657,4 +680,100 @@ TEST_FUNCTION(umock_c_negative_tests_call_count_when_the_module_is_not_initializ
     ASSERT_ARE_EQUAL(size_t, 0, umock_c_indicate_error_call_count);
 }
 
+
+TEST_FUNCTION(umock_c_negative_tests_fail_call_calls_the_call_recorder_fail_call_index_0_result_0)
+{
```
Comment (dcristoloveanu) [inline tests/umock_c_negt_ut/umock_c_negative_tests_ut.c]: Missing Tests_

Comment (dcristoloveanu) [reply tests/umock_c_negt_ut/umock_c_negative_tests_ut.c]: We're doing tagging on these repos (umock_c, uamqp, c shared - I hope :-))

---
In reply to: [224578267](https://github.com/Azure/umock-c/pull/44#discussion_r224578267) [](ancestors = 224578267)

File: tests/umock_c_negt_noini_ut/umock_c_negt_noint_tests.c [hunk 9]
```diff
@@ -337,4 +354,18 @@ TEST_FUNCTION(umock_c_negative_tests_call_count_when_the_module_is_not_initializ
     ASSERT_ARE_EQUAL(size_t, 0, umock_c_indicate_error_call_count);
 }
 
+TEST_FUNCTION(umock_c_negative_tests_can_call_fail_when_the_module_is_not_initialized_returns_default_of_1)
```
Comment (dcristoloveanu) [inline tests/umock_c_negt_noini_ut/umock_c_negt_noint_tests.c]: Missing Tests

File: tests/umock_c_negt_int/umock_c_negt_int_tests.c [hunk 10]
```diff
@@ -426,4 +426,27 @@ TEST_FUNCTION(SetFailReturn_suppresses_paired_calls_tracking_for_mockable_functi
     // no explicit assert, no leak expected
 }
 
+TEST_FUNCTION(umock_c_negative_tests_can_call_fail_test)
+{
```
Comment (dcristoloveanu) [inline tests/umock_c_negt_int/umock_c_negt_int_tests.c]: Missing tagging

File: tests/umockcallrecorder_ut/umockcallrecorder_ut.c [hunk 11]
```diff
@@ -1952,4 +1973,153 @@ TEST_FUNCTION(when_umockcall_set_fail_call_fails_then_umockcallrecorder_fail_cal
     umockcallrecorder_destroy(call_recorder);
 }
 
+/* umockcallrecorder_can_call_fail */
+
+
+/* Tests_SRS_UMOCKCALLRECORDER_31_059: [ umockcallrecorder_can_call_fail shall return in the can_call_fail argument whether the call can fail or not. ]*/
```
Comment (dcristoloveanu) [inline tests/umockcallrecorder_ut/umockcallrecorder_ut.c]: Nitpick: extra line

File: src/umock_c_negative_tests.c [hunk 12]
```diff
@@ -198,3 +198,40 @@ size_t umock_c_negative_tests_call_count(void)
 
     return result;
 }
+
+
```
Comment (dcristoloveanu) [inline src/umock_c_negative_tests.c]: Nitpick: extra line

File: src/umock_c_negative_tests.c [hunk 13]
```diff
@@ -198,3 +198,40 @@ size_t umock_c_negative_tests_call_count(void)
...
+        /* Codes_SRS_UMOCK_C_NEGATIVE_TESTS_31_026: [ If the module was not previously initialized, umock_c_negative_tests_can_call_fail shall return 0. ]*/
+        UMOCK_LOG("umock_c_negative_tests_fail_call: Not initialized.");
+        umock_c_indicate_error(UMOCK_C_ERROR);
+    }
+    else
+    {
+        UMOCKCALLRECORDER_HANDLE call_recorder = umock_c_get_call_recorder();
+        if (call_recorder == NULL)
+        {
+            /* Codes_SRS_UMOCK_C_NEGATIVE_TESTS_31_027: [ If umock_c_get_call_recorder fails, umock_c_negative_tests_can_call_fail shall indicate the error via the umock error callback with error code UMOCK_C_ERROR. ]*/
```
Comment (dcristoloveanu) [inline src/umock_c_negative_tests.c]: > /* Codes_SR [](start = 11, length = 12)

nice

File: src/umockcallrecorder.c [hunk 14]
```diff
@@ -633,3 +633,40 @@ int umockcallrecorder_fail_call(UMOCKCALLRECORDER_HANDLE umock_call_recorder, si
...
+{
+    int result;
+
+    /* Codes_SRS_UMOCKCALLRECORDER_31_056: [ If umock_call_recorder or call_can_fail is NULL, umockcallrecorder_can_call_fail shall return a non-zero value. ]*/
+    if ((umock_call_recorder == NULL) ||
+        (can_call_fail == NULL) ||
+        (index >= umock_call_recorder->expected_call_count))
+    {
+        /* Codes_SRS_UMOCKCALLRECORDER_31_057: [ If index is invalid, umockcallrecorder_can_call_fail shall return a non-zero value. ]*/
+        result = __LINE__;
```
Comment (dcristoloveanu) [inline src/umockcallrecorder.c]: I'd place this req tag between ln 643 and 644

**Comment by dcristoloveanu** [comment]
No major things, some req ids that need to be added in the tests, otherwise looks good to me. I'll have a quick look on the new iteration when it is out and then you can merge it in.

---
In reply to: [429055135](https://github.com/Azure/umock-c/pull/44#issuecomment-429055135) [](ancestors = 429055135)

File: devdoc/umockcallrecorder_requirements.md (hunk 3 above)
Comment (jspaith) [reply devdoc/umockcallrecorder_requirements.md]: Also I used comment #57 twice.  Fixed too.

---
In reply to: [224577211](https://github.com/Azure/umock-c/pull/44#discussion_r224577211) [](ancestors = 224577211)

File: devdoc/umock_c_lib_requirements_with_ids.md (hunk 6 above)
Comment (jspaith) [reply devdoc/umock_c_lib_requirements_with_ids.md]: Could you be more specific on this one?  The way its rendering for me it looks like its following convention (the line 896 call_cannot_fail has two spaces at end of it, which isn't needed i guess, but line 897 I'm seeing a return and I'm not seeing spaces on other places in doc).

---
In reply to: [224577802](https://github.com/Azure/umock-c/pull/44#discussion_r224577802) [](ancestors = 224577802)

Comment (dcristoloveanu) [reply devdoc/umock_c_lib_requirements_with_ids.md]: Oh yes, my bad. leave it as is.

---
In reply to: [224880759](https://github.com/Azure/umock-c/pull/44#discussion_r224880759) [](ancestors = 224880759,224577802)

**Pull request closed without merging**